R/spread_with_duplicates.R

```diff
@@ -0,0 +1,76 @@
+#' Spread a key-value pair across multiple columns in presence of duplicates
+#' 
+#' @param .data A dataframe.
+#' @param key,value Column names or positions.
+#' @param fill If set, missing values will be replaced with this value.
+#' @param aggfunc Aggregation function. 
+#' Default: NA (keep dulicates)
+#' @return A data.frame
+#' @examples 
+#' test1 <- data.frame(col1 = c(1, 1, 1, 1),	
+#'                 col2 = c("H", "H", "H", "H"),
+#'                 key = c("A", "B", "C", "C"),
+#'                 value = c("R", "S", "T", "X"),
+#'                 stringsAsFactors = FALSE)
+#' spread_with_duplicates(test1, key, value)
+#' spread_with_duplicates(test1, 3, 4)
+#' spread_with_duplicates(test1, -2, -1)
+#' 
+#' # with NAs
+#' test2 <- data.frame(col1 = c(1, 1, 1, 2),	
+#'                 key = c("A", "C", "C", "A"),
+#'                 value = c("R", "T", "X", "R"),
+#'                 stringsAsFactors = FALSE)
+#' spread_with_duplicates(test2, key, value)
+#' spread_with_duplicates(test2, key, value, fill = "No idea")
+#' 
+#' # apply aggregate function
+#' test3 <- data.frame(col1 = c(1,1,1,1),
+#'   col2  = c("H", "H", "H", "H"),
+#'   key = c("A", "B", "C", "C"),
+#'   value = c(2, 3, 1, 8),
+#'   stringsAsFactors = FALSE)
+#' spread_with_duplicates(test1, key, value, aggfunc = str_c, collapse = "-")
+#' spread_with_duplicates(test2, key, value, aggfunc = min)
+#' spread_with_duplicates(test2, key, value, aggfunc = mean)
```

Comment:
There is still an issue with the `aggfunc` calculation. Let me show it with an example:

```
library(tidyr)

stocks <- data.frame(
 time = as.Date('2009-01-01') + 0:9,
 X = rnorm(10, 0, 1),
 Y = rnorm(10, 0, 2),
 Z = rnorm(10, 0, 4)
)
stocksm <- stocks %>% gather(key, value, -time)
stocksm[2, 1] <- as.Date('2009-01-01')  # create a duplicate
stocksm %>% spread_with_duplicates(key, value, aggfunc = mean)
```
The output is not as expected, but:

```
         time           X         Y         Z
1  2009-01-01 -0.09567055 -1.078458 -1.814956
2  2009-01-03 -0.09567055 -1.078458 -1.814956
3  2009-01-04 -0.09567055 -1.078458 -1.814956
4  2009-01-05 -0.09567055 -1.078458 -1.814956
5  2009-01-06 -0.09567055 -1.078458 -1.814956
6  2009-01-07 -0.09567055 -1.078458 -1.814956
7  2009-01-08 -0.09567055 -1.078458 -1.814956
8  2009-01-09 -0.09567055 -1.078458 -1.814956
9  2009-01-10 -0.09567055 -1.078458 -1.814956
10 2009-01-02          NA -1.078458 -1.814956
```

Reply:
I would also add a unit test to cover this kind of situation.

Reply:
I agree.. I will work on it